### PR TITLE
wavefront ruby metric lib

### PR DESCRIPTION
**Pull request by yogeshprasad**
It collects the counter, gauge, and histogram from ruby code and sends data to the wavefront.

### CODE REVIEW

File: meters/metric.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+module Meters
```
Comment (sushantdewan123) [inline meters/metric.rb]: can you not call it
call it 'module Measurement'

File: histogram/histogram.rb [hunk 2]
```diff
@@ -0,0 +1,197 @@
+# Metric which calculates the distribution of a value
+#
+require 'tdigest'
+
+require_relative '../meters/metric'
+
+module Meters
```
Comment (sushantdewan123) [inline histogram/histogram.rb]: can you not call it
call it 'module Measurement'

File: meters/counter.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
+require 'concurrent'
+require_relative 'metric'
+
+module Meters
```
Comment (sushantdewan123) [inline meters/counter.rb]: can you not call it
call it 'module Measurement'

File: meters/counter.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        raise RangeError, "Counter needs an integer. Given #{val}"
+      end
+
+      @value.increment(val)
+    end
+
+    # Decrement counter by val (default is 1)
+    #
+    # @param val [Integer] Counter decrement by val
+    def dec(val = 1)
```
Comment (sushantdewan123) [inline meters/counter.rb]: let's not even support dec method

File: meters/counter.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+      unless val.is_a? Integer
+        raise RangeError, "Counter needs an integer. Given #{val}"
+      end
+
+      @value.increment(-val)
+    end
+
+    # Reset the counter to 0
+    def reset
```
Comment (sushantdewan123) [inline meters/counter.rb]: let's not even support reset method.

File: registry.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+                      clock_func = nil)
+      add(Meters::Histogram.new(name, point_tags, accuracy, granularity, max_bins, clock_func))
+    end
+
+    # Check if metric exists
+    #
+    # @param name [String] name of the metric
+    # @return [Bool]
+    def exist?(name)
+      @store.key?(name.to_sym)
```
Comment (sushantdewan123) [inline registry.rb]: You can have 2 counters with the same name
but different point tags
those 2 should not be treated as duplicates.

**Review by sushantdewan123 — CHANGES_REQUESTED** [review]

File: registry.rb [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    end
+
+    # Add metric into registry
+    #
+    # @param metric [Metric] metric to add it can be of any type
+    # @return metric object
+    def add(metric)
+      raise TypeError unless metric.respond_to? :name
+
+      key = metric.name
```
Comment (sushantdewan123) [inline registry.rb]: this is incorrect
see 

https://github.com/wavefrontHQ/wavefront-pyformance/blob/73b7a43699ee3b4a2e21bf251630b473f27f2d08/wavefront_pyformance/tagged_registry.py#L15

https://github.com/wavefrontHQ/go-metrics-wavefront/blob/e86b11d2e81bbf4e48f1fde1fe8ad5e4b8b5e40e/reporting/utils.go#L10
^^ is used here: https://github.com/wavefrontHQ/go-metrics-wavefront/blob/e86b11d2e81bbf4e48f1fde1fe8ad5e4b8b5e40e/reporting/registry.go#L20

**Review by sushantdewan123 — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -1,2 +1,87 @@
...
+## Core Features
+### Counter
+Counters show information over time. Think of a person with a counter at the entrance to a concert. The counter shows the total number of people that have entered so far. Counters usually increase over time but might briefly go to zero.
+### Gauge
+A gauge shows the current value for each point in time. Think of a thermometer that shows the current temperature or a gauge that shows how much electricity your Tesla has left.
+
+### Histogram
+a mechanism to compute, store, and use distributions of metrics.
+
+## Examples
```
Comment (mshvol) [inline README.md:21]: Add ` ```ruby` code tag to all code snippets.

File: README.md [hunk 9]
```diff
@@ -1,2 +1,87 @@
...
+    
+    store = Registry::MetricsRegistry.new
+    test_gauge = store.gauge("test.gauge", {"tag1"=>"value1", "tag2"=>"value2"}, 0)
+    test_gauge.set(10)
+    
+    puts test_gauge.value         # 10.0
+    puts test_gauge.name          # test.gauge
+    puts test_gauge.point_tags    # {"tag1"=>"value1", "tag2"=>"value2"}
+    
+#### Histogram
```
Comment (mshvol) [inline README.md:53]: Example has to be improved. The `clock` func is mostly for testing. It is rarely, if ever, used. Even max_bins isn't generally changed. The most common case is ignore all optional params.

File: README.md [hunk 10]
```diff
@@ -1,2 +1,87 @@
...
+
+    client = Wavefront::WavefrontDirectIngestionClient.new(server, token)
+    store = Registry::MetricsRegistry.new
+    Reporters::Wavefront.new(client, store).start
+    
+    $test_count = store.counter("test.call", {"tag1"=>"value1", "tag2"=>"value2"}, 0)
+    5.times do
+        test
+    end
+
```
Comment (mshvol) [inline README.md]: Maybe add `reporter.stop()` at the end to show good practice?

File: meters/metric.rb [hunk 11]
```diff
@@ -0,0 +1,13 @@
+module Measurement
+  class Metric
+    attr_reader :name, :point_tags
+
+    # @param name [String] metric name
+    # @param point_tags [Hash] metric point tags
+    def initialize(name, point_tags)
+      # TODO: sanitize/validate both?
+      @name = name.to_sym
+      @point_tags = point_tags&.clone || {}
```
Comment (mshvol) [inline meters/metric.rb]: Now that we are also using `point_tags` for keys, use `.freeze` so that point_tags and hash key are always consistent. I don't think this'll have an issue with `sender.send_metric()` but maybe test with freeze just in case.

File: registry.rb [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        raise DuplicateKeyError, "Cannot add duplicate metric into registry: #{key}"
+      end
+
+      metric
+    end
+
+    # Remove the metric from registry
+    #
+    # @param name [String] name of the metric
+    def del(name, point_tags)
```
Comment (mshvol) [inline registry.rb]: nitpick: `point_tags` could be made optional `={}`?

File: registry.rb [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    def metrics
+      @store.values
+    end
+
+    # Encode the key using point tags to allow a metric with different point tags
+    #
+    # @param key [String] name of the metric
+    # @param tags [Hash] list of metric point tags
+    # @return [String] encoded key
+    def encode_key(key, tags)
```
Comment (mshvol) [inline registry.rb]: Any reason we are mixing in `Object`s as the `@store` keys if tags are empty (In our case the `Object`s will be `Symbol`)? i think it is better if we have consistently one type in the Hash. Something like this-
```ruby
def encode_key(key, tags)
    key = key.to_s
    if !tags.nil? && !tags.empty?
        key += "-tags=" + tags.to_json
    end
    return key
end
```
Maybe add test case for key with empty `point_tags` as well?

**Review by mshvol — CHANGES_REQUESTED** [review]
Minor suggestions.

File: reporters/reporter.rb [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        end
+      end
+    end
+
+    # Stop reporting
+    def stop
+      # Flush all buffer before close the client.
+      @lock.synchronize do
+        @closed = true
+        @task.join
```
Comment (mshvol) [inline reporters/reporter.rb]: @yogeshprasad While this statement is logically correct (and I understand I suggested this approach), this could cause a lot of waiting when `reporting_interval` is set to a large value. I realized this when I was writing code for  heartbeat service and I think I have a better approach. I'll share it tomorrow. Apart from that, this looks fine.

The latest commit also includes some changes needed for my work.

**Comment by mshvol** [comment]
Just a reminder comment. After merging please be sure to change the travis label in README to point to this repo.

**Review by mshvol — APPROVED** [review]
LGTM

File: README.md [hunk 15]
```diff
@@ -1,6 +1,13 @@
 # wavefront-ruby-metrics [![Build Status](https://travis-ci.com/yogeshprasad/wavefront-ruby-metrics.svg?branch=master)](https://travis-ci.com/yogeshprasad/wavefront-ruby-metrics)
 
-It collects the counter, gauge, and histogram from your ruby code
+It collects the counter, gauge, and histogram from your ruby code.
```
Comment (mshvol) [inline README.md]: Could be described better? Something like-
> Captures application-level metrics (counters, gauges & histograms) for your ruby code.
> Supports reporting to Wavefront.

File: lib/wavefront/metrics/reporters/wavefront.rb [hunk 16]
```diff
@@ -25,12 +25,13 @@ def initialize(sender, registry = nil, reporting_interval = 5, host = Socket.get
     def report_now
       @registry.metrics.each do |data|
         if data.class == Measurement::Counter or data.class == Measurement::Gauge
-          @sender.send_metric(data.name, data.value, nil, @host, data.point_tags)
+          result = @registry.get_metrics(data)
+          @sender.send_metric(data.name.to_s + "." + result.keys[0].to_s, data.value, nil, @host, data.point_tags)
```
Comment (mshvol) [inline lib/wavefront/metrics/reporters/wavefront.rb]: Since you already have a general function now to support multiple metric "fields" in one metric, maybe add an if for wavefront histograms and for everything else use a loop like this? _(As everything else goes through_ `send_metric` _anyway in this reporter)_

```ruby
@registry.get_metrics(data).each do |k,v|
  @sender.send_metric(data.name.to_s + "." + k.to_s, v, nil, @host, data.point_tags)
end
```

File: lib/wavefront/metrics/reporters/reporter.rb [hunk 17]
```diff
@@ -35,15 +35,15 @@ def stop
         begin  
           report_now
         rescue Exception => e
-          puts "Reporter Exception: #{e.inspect}"
+          raise e
```
Comment (mshvol) [inline lib/wavefront/metrics/reporters/reporter.rb]: **Potential issue**: Wouldn't this cause the reporter thread to stop running on the first exception? This won't be detected in the main thread until this thread is joined (which is at `stop()`). Any reason this was changed?

File: lib/wavefront/metrics/registry.rb [hunk 18]
```diff
@@ -96,5 +96,19 @@ def encode_key(key, tags)
       end
       return key
     end
+
+    # Return the metric value with suffix
+    #
+    # @param metric [Metric] metric object
+    # @return [Hash] metric list
+    def get_metrics(metric)
```
Comment (mshvol) [inline lib/wavefront/metrics/registry.rb]: Maybe rename to `get_metric_fields`? Similar to telegraf terminology and such

**Review by mshvol — COMMENTED** [review]
Added a few suggestions. Most of them are nitpicks. Feel free to ignore.

File: lib/wavefront/metrics/reporters/wavefront.rb (hunk 16 above)
Comment (yogeshprasad) [reply lib/wavefront/metrics/reporters/wavefront.rb]: in the case of counter and gauges, I think a loop is not required as we know there will be only one value at a time

**Review by mshvol — APPROVED** [review]
New changes LGTM

File: lib/wavefront/metrics/reporters/wavefront.rb [hunk 19]
```diff
@@ -28,7 +28,7 @@ def report_now
         if data.class == Measurement::Counter or data.class == Measurement::Gauge
           result = @registry.get_metric_fields(data)
           @sender.send_metric(data.name.to_s + "." + result.keys[0].to_s, data.value,
-                              nil, @host, data.point_tags.merge(@application_tags))
+            Time.now.to_i, @host, data.point_tags.merge(@application_tags))
```
Comment (mshvol) [inline lib/wavefront/metrics/reporters/wavefront.rb]: This returns time in seconds. To be consistent, we need milliseconds. Use 
```ruby
(Time.now.to_f * 1000).round
```

File: README.md [hunk 20]
```diff
@@ -1,2 +1,94 @@
...
+```
+
+## Core Features
+### Counter
+Counters show information over time. Think of a person with a counter at the entrance to a concert. The counter shows the total number of people that have entered so far. Counters usually increase over time but might briefly go to zero.
+### Gauge
+A gauge shows the current value for each point in time. Think of a thermometer that shows the current temperature or a gauge that shows how much electricity your Tesla has left.
+
+### Histogram
+a mechanism to compute, store, and use distributions of metrics.
```
Comment (sushantdewan123) [inline README.md:19]: Capital `A`

**Review by sushantdewan123 — APPROVED** [review]
